### PR TITLE
Fix volume transformation, add observation wrapper

### LINKED ISSUE #8: Volumes interpolation
All volumes should have the same spacing, hence resolution.
The size of the volumes should be changed accordingly, so that:
`new_size = physical_size / spacing`

### LINKED ISSUE #9: Relative volume direction (orientation) considered in slicing actions
Since the volume orientation is saved under GetDirection(), the random rotation applied do augment the data does not have any effect on the slicing actions. In fact, it is possible to get the same slice without changing the action rotation, even though the volume has been transformed with a random rotation.

### PR DESCRIPTION
**Pull request by carlocagnetta**
- Fix observations to MultiBoxSpace:
  All observations_spaces are now MultiBox
- Fix volume transformation:
  Was buggy when transformations were laying the volume origin in negative absolute coordinates, 
  Solution: project the origin into positive octant
- Fix clustering:
  Buggy behavior when no noise cluster was present. Fixed that as well as parameters.
- New volumes src
  New src folder for volumes' operations
- AddRewardDetailsWrapper:
  Wrapper to keep memory of the best reward and action seen so far.
  Best reward and action are stacked at the end of every observation.
- Add tests for optimal slice
- Add more volumes
- Normalizing volumes to same spacing
- Add class for ImageVolume and TransformedVolume

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -154,7 +154,7 @@ lint = ["_black_check", "_ruff_check", "_ruff_check_nb"]
 _poetry_install_sort_plugin = "poetry self add poetry-plugin-sort"
 _poetry_sort = "poetry sort"
 clean-nbs = "python docs/nbstripout.py"
-format = ["_black_format", "_ruff_format", "_ruff_format_nb", "_poetry_install_sort_plugin", "_poetry_sort"]
```
Comment (carlocagnetta) [inline pyproject.toml:157]: this is a mistake. I removed this only so that it would run on the VM. did not mean to commit it

File: scripts/random_volume_transformations.ipynb [hunk 2]
```diff
@@ -0,0 +1,190 @@
+{
```
Comment (MischaPanch) [inline scripts/random_volume_transformations.ipynb]: Is in in scripts instead of docs on purpose? You can add a separate dir called notebooks for non-documentation notebooks

File: src/armscan_env/envs/labelmaps_navigation.py [hunk 3]
```diff
@@ -168,8 +168,8 @@ def get_optimal_action_array(self) -> np.ndarray:
         full_action_arr = self.get_full_optimal_action_array()
         return full_action_arr[self._get_projected_action_arr_idx()]
 
-    def step_to_optimal_state(self) -> None:
-        self.step(self.get_optimal_action())
+    def step_to_optimal_state(self) -> tuple[TObs, float, bool, bool, dict[str, Any]]:
```
Comment (MischaPanch) [inline src/armscan_env/envs/labelmaps_navigation.py]: TObs -> Observation[LabelmapStateAction, Any]

File: src/armscan_env/envs/labelmaps_navigation.py [hunk 4]
```diff
@@ -272,32 +269,22 @@ def apply_volume_transformation(
...
-            rotation=(
-                optimal_action.rotation[0] + small_random_z_rotation,
-                optimal_action.rotation[1] + small_random_x_rotation,
-            ),
-            translation=(
-                optimal_action.translation[0],
-                optimal_action.translation[1] + small_random_y_translation,
-            ),
+        volume_transformation = ManipulatorAction(
+            rotation=(np.random.uniform(-20, 20), np.random.uniform(-5, 5)),
```
Comment (MischaPanch) [inline src/armscan_env/envs/labelmaps_navigation.py]: the transformation and the sampling should be separated. This function should receive a sampled rotation and translation instead of sampling here

File: src/armscan_env/envs/labelmaps_navigation.py [hunk 5]
```diff
@@ -272,32 +269,22 @@ def apply_volume_transformation(
...
-        transformed_optimal_action = ManipulatorAction(
-            rotation=(
-                optimal_action.rotation[0] + small_random_z_rotation,
-                optimal_action.rotation[1] + small_random_x_rotation,
-            ),
-            translation=(
-                optimal_action.translation[0],
-                optimal_action.translation[1] + small_random_y_translation,
-            ),
+        volume_transformation = ManipulatorAction(
```
Comment (MischaPanch) [inline src/armscan_env/envs/labelmaps_navigation.py]: This function is complicated enough to deserve a docstring ;)

File: src/armscan_env/envs/state_action.py [hunk 6]
```diff
@@ -29,6 +30,9 @@ def to_normalized_array(
         # normalize translation to [-1, 1]: 0 -> -1, translation_bounds -> 1
         rotation = np.zeros(2)
         translation = np.zeros(2)
+        if self.translation[0] < 0 or self.translation[1] < 0:
```
Comment (MischaPanch) [inline src/armscan_env/envs/state_action.py]: when would that happen? debug logging seems a bit weak, can we raise an error? Actions should always be valid, I feel

File: src/armscan_env/envs/state_action.py [hunk 7]
```diff
@@ -43,7 +47,7 @@ def to_normalized_array(
         result = np.concatenate([rotation, translation])
 
         if not (result >= -1).all() or not (result <= 1).all():
-            raise ValueError(
```
Comment (MischaPanch) [inline src/armscan_env/envs/state_action.py:46]: why this change?

Comment (MischaPanch) [inline src/armscan_env/envs/state_action.py:46]: why this change?

File: src/armscan_env/envs/state_action.py (hunk 6 above)
Comment (MischaPanch) [inline src/armscan_env/envs/state_action.py]: when would that happen? debug logging seems a bit weak, can we raise an error? Actions should always be valid, I feel

File: src/armscan_env/volumes/loading.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import SimpleITK as sitk
+
+
+def load_sitk_volume(
```
Comment (MischaPanch) [inline src/armscan_env/volumes/loading.py]: is this now being used everywhere for loading images?

File: src/armscan_env/volumes/slicing.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    # Verify the shapes
+    log.debug(
+        f"Original Array Shape: {original_array.shape}\n"
+        f"Padded Array Shape: {padded_array.shape}",
+    )
+
+    return padded_array
+
+
+class EulerTransform:
```
Comment (MischaPanch) [inline src/armscan_env/volumes/slicing.py]: This should be in the same module as ManipulatorAction, it's not related to slicing

File: src/armscan_env/volumes/slicing.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        log.debug(
+            f"Random transformation: {self.action}\n"
+            f"Original action: {relative_action}\n"
+            f"Transformed action: {transformed_action}\n",
+        )
+
+        return transformed_action
+
+
+class TransformedVolume(sitk.Image):
```
Comment (MischaPanch) [inline src/armscan_env/volumes/slicing.py]: this kind of inheritance can be very problematic. I will push a change (still hacky, but better than this)

File: src/armscan_env/volumes/slicing.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    resampler.SetInterpolator(sitk.sitkNearestNeighbor)
+
+    # Resample the volume on the arbitrary plane
+    transformed_volume = resampler.Execute(volume)
+    # Todo: hack --> needed to break rotation dependency
+    return TransformedVolume(transformed_volume, action)
+
+
+def slice_volume(
+    volume: sitk.Image | TransformedVolume,
```
Comment (MischaPanch) [inline src/armscan_env/volumes/slicing.py]: You should only ever used TransformedVolume instead of bare `sitk.Image` in your code now. The `load_image` function should load an instance of it and there should be a default value for the transform (probably `None` will do)

File: src/armscan_env/wrapper.py [hunk 12]
```diff
@@ -128,26 +164,116 @@ def __getattr__(self, item: str) -> Any:
...
+class AddRewardDetailsWrapper(AddObservationsWrapper):
+    @property
+    def additional_obs_space(self) -> gym.spaces:
+        return self._additional_obs_space
+
+    def __init__(
+        self,
+        env: LabelmapEnv | Env[ObsType, ActType],
+        num_steps_to_observe: int | None = None,
+        additional_obs: Observation = ActionRewardObservation().to_array_observation(),
```
Comment (MischaPanch) [inline src/armscan_env/wrapper.py]: mutable defaults are evil. In the long run we need an observation registry and proper constructors. For now, pls just add a todo about adding that registry

File: src/armscan_env/wrapper.py [hunk 13]
```diff
@@ -128,26 +164,116 @@ def __getattr__(self, item: str) -> Any:
...
+                f"Observation spaces are not of type Box: {type(self.env.observation_space)}, {type(self.additional_obs_space)}",
+            )
+
+    @property
+    @abstractmethod
+    def additional_obs_space(self) -> gym.spaces:
+        pass
+
+    @abstractmethod
+    def get_additional_obs(
```
Comment (MischaPanch) [inline src/armscan_env/wrapper.py]: rename to `get_additional_obs_array` pls to disambiguate with existing field `additional_obs`

File: src/armscan_env/wrapper.py [hunk 14]
```diff
@@ -128,26 +164,116 @@ def __getattr__(self, item: str) -> Any:
...
+        )
+        clustering_reward = anatomy_based_rwd(tissue_clusters=tissue_clusters)
+        self.rewards.append(clustering_reward)
+        self.states.append(self.env.cur_state_action)
+        self.highest_rew_state_arr = self.states[np.argmax(self.rewards)]
+
+    def get_additional_obs(self) -> np.ndarray:
+        # base_obs is not used, instead we directly access the current image from the env
+        self._update_observation_fields()
+        return self.additional_obs.compute_observation(self.highest_rew_state_arr)
```
Comment (MischaPanch) [inline src/armscan_env/wrapper.py:300]: Doesn't this mean the reward is recomputed again, in addition to line 265? Also, can't we extract the envs reward instead of computing it in the wrapper?

File: src/armscan_env/wrapper.py (hunk 12 above)
Comment (MischaPanch) [reply src/armscan_env/wrapper.py]: I feel like this shouldn't be passed here at all. This wrapper is supposed to add reward details, and in fact no other additional_obs can be passed (it would break your code below). Would also solve the registry problem, no need for a todo then

File: src/armscan_env/wrapper.py [hunk 15]
```diff
@@ -216,35 +363,13 @@ def create_env(self, mode: EnvMode) -> LabelmapEnv:
             apply_volume_transformation=self.apply_volume_transformation,
         )
 
+        if self.add_reward_details:
```
Comment (MischaPanch) [inline src/armscan_env/wrapper.py:396]: this should happen after frame-stack, not before. **Important:** changing this will invalidate your trainings

File: test/armscan_env/test_labelmap_volumes.py [hunk 16]
```diff
@@ -5,7 +5,8 @@
 import SimpleITK as sitk
 from armscan_env.clustering import TissueLabel
 from armscan_env.config import get_config
-from armscan_env.slicing import slice_volume
+from armscan_env.envs.state_action import ManipulatorAction
+from armscan_env.volumes.slicing import slice_volume
```
Comment (MischaPanch) [inline test/armscan_env/test_labelmap_volumes.py:12]: I feel like some tests on the correctness of the transformations can be added. After addresssing the comment on separating sampling and transformation, you can actually write deterministic tests

**Review by MischaPanch — CHANGES_REQUESTED** [review]

**Comment by MischaPanch** [comment]
Pls also fix the failing pipeline - why do the linting tests not start?

**Comment by MischaPanch** [comment]
Pls also fix the failing pipeline - why do the linting tests not start?

**Comment by MischaPanch** [comment]
Pls also fix the failing pipeline - why do the linting tests not start?

File: src/armscan_env/volumes/slicing.py [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+    :param volume: 3D volume to be sliced
+    :param slice_shape: shape of the output slice
+    :param action: action to transform the volume
+    :return: the sliced volume.
+    """
+    o = np.array(volume.GetOrigin())
+
+    if isinstance(volume, TransformedVolume):
+        # TODO: why is origin not used here?
```
Comment (MischaPanch) [inline src/armscan_env/volumes/slicing.py]: @carlocagnetta pls address the TODOs in this function :)

File: docs/02_notebooks/L2_DBSCAN_clustering.ipynb
Comment (carlocagnetta) [inline docs/02_notebooks/L2_DBSCAN_clustering.ipynb:1]: @MischaPanch might you have forgotten to clean-nbs?

Comment (MischaPanch) [inline docs/02_notebooks/L2_DBSCAN_clustering.ipynb:1]: Done and force pushed, sry for that

File: src/armscan_env/envs/labelmaps_navigation.py (hunk 3 above)
Comment (carlocagnetta) [reply src/armscan_env/envs/labelmaps_navigation.py]: I have the same return as in the step function, is this not correct?

File: src/armscan_env/envs/state_action.py (hunk 6 above)
Comment (carlocagnetta) [reply src/armscan_env/envs/state_action.py]: it happens when the transformation is negative, and transforms the optimal action into negative octant. Reason for which it is necessary to reproject the optimal action into positive octant.
In the end it just defines a coordinate system of a plane, so projecting it still defines the same plane.

File: src/armscan_env/envs/labelmaps_navigation.py (hunk 4 above)
Comment (carlocagnetta) [reply src/armscan_env/envs/labelmaps_navigation.py]: where do you see the action sampling function, in the env or in the ManipulatorAction class? Should the range be configurable?

File: src/armscan_env/wrapper.py (hunk 15 above)
Comment (carlocagnetta) [reply src/armscan_env/wrapper.py:396]: It is not possible to do it after framestack. The additional observation does not have the same size as the stacked observations! I think it can only be done like this. I don't understand how you would add the reward information otherwise.

File: src/armscan_env/envs/labelmaps_navigation.py (hunk 3 above)
Comment (MischaPanch) [reply src/armscan_env/envs/labelmaps_navigation.py]: It is not, because step is defined inside the generic base class, whereas this method is defined in the non-generic implementation

File: src/armscan_env/volumes/slicing.py (hunk 17 above)
Comment (carlocagnetta) [reply src/armscan_env/volumes/slicing.py]: this whole thing is a bit of a brain teaser actually.

File: test/armscan_env/test_labelmap_volumes.py (hunk 16 above)
Comment (carlocagnetta) [reply test/armscan_env/test_labelmap_volumes.py:12]: there is already an issue for that, and it is in the todos

File: src/armscan_env/envs/labelmaps_navigation.py (hunk 4 above)
Comment (MischaPanch) [reply src/armscan_env/envs/labelmaps_navigation.py]: The rotation and translation are sampled here. This is not good design (violated open closed and single responsibility principle). Instead, the function should take a rotation and translation as args and sampling should happen in downstream code

File: src/armscan_env/wrapper.py (hunk 14 above)
Comment (carlocagnetta) [reply src/armscan_env/wrapper.py:300]: We need to recompute the reward, as we do also in the observations, bc otherwise the last reward is taken. the observations are computed before updating the reward if I am correct.

File: src/armscan_env/wrapper.py (hunk 15 above)
Comment (MischaPanch) [reply src/armscan_env/wrapper.py:396]: Let's have a call tomorrow, it really should not happen before stacking, that doesn't make much sense and it unnecessarily enlarges the obs space

File: src/armscan_env/wrapper.py (hunk 12 above)
Comment (carlocagnetta) [reply src/armscan_env/wrapper.py]: But then how do I set the additional observation space? since it depends on the action_space.shape

File: src/armscan_env/volumes/slicing.py (hunk 10 above)
Comment (MischaPanch) [reply src/armscan_env/volumes/slicing.py]: I stuck to the current interface in the end and just added some "safety" features

File: src/armscan_env/volumes/slicing.py (hunk 11 above)
Comment (MischaPanch) [reply src/armscan_env/volumes/slicing.py]: Obsolete after my last commits

File: src/armscan_env/wrapper.py (hunk 14 above)
Comment (MischaPanch) [reply src/armscan_env/wrapper.py:300]: Yeah, that design was a footgun, but not easily fixable now. Let's recompute, maybe adding @cache in the appropriate place will solve performance issues

File: src/armscan_env/wrapper.py (hunk 12 above)
Comment (MischaPanch) [reply src/armscan_env/wrapper.py]: In the call to super init. No need to let the user pass it

File: src/armscan_env/volumes/loading.py (hunk 8 above)
Comment (carlocagnetta) [reply src/armscan_env/volumes/loading.py]: not yet

File: src/armscan_env/envs/state_action.py (hunk 7 above)
Comment (carlocagnetta) [reply src/armscan_env/envs/state_action.py:46]: because this might happen, and it is dealt with it. Per see, the action is still valid, it is just not possible to get such an action from the agent
